Mipmaps

mattigruener opened the pull request:
Hey guys,

could you have a look? This is so that we can load particular miplevels from textures for GL. It's a little fiddly...

Thanks, 

Matti.

mattigruener commented on src/IECoreGL/TextureLoader.cpp:
```diff
@@ -37,19 +37,27 @@
 #include "IECoreGL/Texture.h"
 #include "IECoreGL/ToGLTextureConverter.h"
 
+#include "IECoreImage/ImageReader.h"
```
Is this dependency acceptable?

johnhaddon replied:
Yep. IECoreGL has a dependency on IECoreImage already.

johnhaddon commented on include/IECoreGL/TextureLoader.h:
```diff
@@ -62,7 +62,7 @@ class IECOREGL_API TextureLoader : public IECore::RefCounted
 		TextureLoader( const IECore::SearchPath &searchPaths );
 
-		TexturePtr load( const std::string &name );
+		TexturePtr load( const std::string &name, const IECore::IntData *maximumResolution = nullptr );
```
`IntData` seems like overkill here? Anything wrong with `int`? We could just use `numeric_limits<int>::max` as the default.

Should we be concerned about the breaking change at this juncture? I don't mind for Gaffer's purposes, but it could cause problems at IE and elsewhere perhaps? If so, we could just add an overload with the extra argument to maintain compatibility, and have the old one forward to the new one.

mattigruener replied:
Addressed in 2ae7040 and 164a72a. Sorry, kept them separate to make squashing easier later.

johnhaddon commented on src/IECoreGL/TextureLoader.cpp:
```diff
@@ -37,19 +37,27 @@
-	TexturesMap::const_iterator it = m_loadedTextures.find( name );
+	int dataWindowSizeLimit = maximumResolution ? maximumResolution->readable() : std::numeric_limits<int>::max();
+
+	TexturesMapKey key = std::make_pair( name, dataWindowSizeLimit );
```
As I understand it, the main purpose of `make_pair` is to avoid typing out the types, and for generic code where determinging the type might be a pain. Here `TexturesMapKey key( name, dataWindowSizeLimit )` is shorter and clearer.

mattigruener replied:
Fixed as part of 2ae7040. Thanks, you're completely right, of course.

johnhaddon commented on src/IECoreGL/TextureLoader.cpp:
```diff
@@ -59,24 +67,49 @@ TexturePtr TextureLoader::load( const std::string &name )
+		return nullptr;
+	}
+
+	IECoreImage::ImageReaderPtr imageReader = IECore::runTimeCast<IECoreImage::ImageReader>( r );
```
If we know we can only use an ImageReader, then using the generic `Reader::create()` is a bit misleading. What about replacing the conditional above with `ImageReader::canRead( fileName )`, and then just doing `new ImageReader( fileName )` here?

mattigruener replied:
How's this? 26fc68c

johnhaddon commented on test/IECoreGL/TextureLoaderTest.py:
```diff
@@ -56,6 +56,32 @@ def testLoadGreyscale( self ) :
+		size = i.dataWindow.size()
+
+		# Test if resolution is adhering to our max resolution and has a reasonable size.
+		self.assertTrue( max( size.x, size.y ) <= maxResolution )
```
`assertLessEqual()` and `assertGreater()` give better diagnostics when they fail - it's worth getting into the habit of using them.

mattigruener replied:
Also much better, thanks :) 7ced655

johnhaddon commented on src/IECoreGL/bindings/TextureLoaderBinding.cpp:
```diff
@@ -43,14 +43,31 @@
 		.def( init<const IECore::SearchPath &>() )
-		.def( "load", &TextureLoader::load )
+		.def( "load", &load1 )
+		.def( "load", &load2 )
```
You should be able to achieve the same thing without helpers and with a single call to `def`, at the same time as providing better python help on the function signature. Something like this (untested) :

```
def( "load", &TextureLoader::load, ( arg( "fileName" ), arg( "maximumResolution" ) = numeric_limits<int>::max() ) );
```

mattigruener replied:
Thanks, much better. Addressed in 26c9dcc

johnhaddon commented on include/IECoreImage/ImageReader.h:
```diff
@@ -118,6 +122,7 @@ class IECOREIMAGE_API ImageReader : public IECore::Reader
 
 		IECore::StringVectorParameterPtr m_channelNamesParameter;
 		IECore::BoolParameterPtr m_rawChannelsParameter;
+		IECore::IntParameterPtr m_miplevelParameter;
```
Just noting that if we needed to, we could avoid the breaking change by looking the parameter up by name.

johnhaddon commented on include/IECoreImage/ImageReader.h:
```diff
@@ -81,6 +82,9 @@ class IECOREIMAGE_API ImageReader : public IECore::Reader
 		IECore::BoolParameter *rawChannelsParameter();
 		const IECore::BoolParameter *rawChannelsParameter() const;
+		/// The parameter specifying the miplevel to be read from the image file.
+		IECore::IntParameter *miplevelParameter();
```
`mipLevel` perhaps? Camel case L?

mattigruener replied:
Addressed in dffb0de. Note to myself: This mixes ImageReader and TextureLoader, - needs fixing before squashing.

johnhaddon commented on src/IECoreImage/ImageReader.cpp:
```diff
@@ -429,6 +438,29 @@ class ImageReader::Implementation
+			int result = 0;
+
+			ConstIntParameterPtr p = m_reader->miplevelParameter();
+			const IntData *d = static_cast<const IntData *>( p->getValue() );
```
No need to muck about with data here - `int result = m_reader->miplevelParameter()->getNumericValue()` should do the trick in one line.

mattigruener replied:
Fixed in dffb0de

johnhaddon commented on src/IECoreGL/ShaderStateComponent.cpp:
```diff
@@ -43,12 +43,19 @@
 
+namespace
+{
+	const InternedString g_maxTextureResolutionParameterSuffix( "_maxResolution" );
```
Can we use ":maxResolution" here - that feels like it better matches are general conventions for namespacing things. Since ":" isn't a valid character for a true GLSL shader parameter, I think it eliminates any possibility of a clash.

Also, don't indent for namespaces.

mattigruener replied:
Done in a7d166d

mattigruener commented:
Hey John,

thanks for the feedback :) I've addressed all comments inline.

> I'm afraid I didn't pick up on anything during the review that might account for that. The one thing I did wonder about was whether it was kosher to turn on "automip" part-way through using the cache - I suspect the expected use might be to turn it on before doing anything else. Actually, can't we just turn it on to start with anyway?

Could you take a look at what I've done in b9a917a? I couldn't recreate the crash locally at all - will keep an eye on what Travis thinks of this set of changes.

Regarding breaking changes - I've left them in for now as I need to touch this again anyway to squash things in. I'm not sure if we need to avoid them? I'd usually rely on @andrewkaufman to make that decision, - do you have a preference, John?

I'm keeping the label for now as I still need to do the history cleanup. Hopefully that's in line with the new guidelines.

Thanks,

Matti.

mattigruener commented:
Had to add another commit. Would've squashed it in but I think that would've messed up the commit hashes I used in my inline comments.

johnhaddon commented:
> thanks for the feedback :) I've addressed all comments inline.

Thanks Matti - LGTM.

> Regarding breaking changes...do you have a preference, John?

For my purposes, cleaner code and a breaking change is fine, as long as it's synced with a Gaffer major version. And Cortex 10 isn't final yet, so we're not breaking any promises. I just wanted to highlight it in case it caused issues for IE...

mattigruener commented:
Thanks John! :) I have squashed in the commits that contained fixes. Waiting for checks to complete before merging.

mattigruener commented:
Did you see the test failures before you approved this, John? I'm not sure what's happening there.

johnhaddon commented:
Ugh. I did not. Seems to be this :

```
Traceback (most recent call last):
  File "contrib/IECoreAppleseed/test/IECoreAppleseed/All.py", line 41, in <module>
    from MeshAlgoTest import MeshAlgoTest
  File "/home/travis/build/ImageEngine/cortex/contrib/IECoreAppleseed/test/IECoreAppleseed/MeshAlgoTest.py", line 41, in <module>
    import IECoreScene
  File "/home/travis/build/ImageEngine/cortex/python/IECoreScene/__init__.py", line 37, in <module>
    from _IECoreScene import *
ImportError: No module named _IECoreScene
```

I think somewhere we're not declaring a dependency, so it occasionally tries to run the appleseed tests before it's built the IECoreScene python module.

johnhaddon commented:
Should we merge this one, or are we waiting for Andrew's review and decision on the compatibility?

mattigruener commented:
I think, as we'll most likely want this included in Gaffer 0.54.0.0, we can't wait for Andrew. My understanding is that this is only going to affect internal projects, if they're using either `TextureLoader::load()` or `IECoreImage::ImageReader`. If that's correct, I think only IERendering is going to be affected and will need a recompile. One other project uses `ImageReader` but only through the Python API.

Ivan and me are thinking that we should just go ahead and merge this - potentially checking with Lucien first about his [currently open PR](https://github.com/ImageEngine/cortex/pull/963) to see if we need to release that separately so that we can roll back to a version that only contains that work if we need to revert the changes for some unforeseen reason.

johnhaddon commented:
OK, cool - I'll leave it to you and Ivan to manage a merge/release then. I've just merged Lucien's PR, thinking that is what you wanted, but now I'm worried I may have misinterpreted your comment?
